pkg/deploy/controller/imagechange/factory.go

```diff
@@ -3,82 +3,176 @@ package imagechange
+	for _, stream := range c.streamStore.GetStreamsForConfig(dc) {
+		glog.V(4).Infof("Reconciling stream %q for config %q\n", stream.Name, dc.Name)
+		c.enqueueImageStream(stream)
+	}
+}
+
+func (c *ImageChangeController) enqueueImageStream(stream *imageapi.ImageStream) {
+	key, err := kcontroller.KeyFunc(stream)
+	if err != nil {
+		glog.Errorf("Couldn't get key for object %+v: %v", stream, err)
+		return
+	}
+	c.queue.Add(key)
+}
+
+func (c *ImageChangeController) worker() {
+	for {
+		if quit := c.work(); quit {
+			return
+		}
+	}
 }
 
-// Create creates an ImageChangeController.
-func (factory *ImageChangeControllerFactory) Create() controller.RunnableController {
-	imageStreamLW := &cache.ListWatch{
-		ListFunc: func(options kapi.ListOptions) (runtime.Object, error) {
-			return factory.Client.ImageStreams(kapi.NamespaceAll).List(options)
-		},
-		WatchFunc: func(options kapi.ListOptions) (watch.Interface, error) {
-			return factory.Client.ImageStreams(kapi.NamespaceAll).Watch(options)
-		},
-	}
-	queue := cache.NewResyncableFIFO(cache.MetaNamespaceKeyFunc)
-	cache.NewReflector(imageStreamLW, &imageapi.ImageStream{}, queue, 2*time.Minute).Run()
-
-	deploymentConfigLW := &cache.ListWatch{
-		ListFunc: func(options kapi.ListOptions) (runtime.Object, error) {
-			return factory.Client.DeploymentConfigs(kapi.NamespaceAll).List(options)
-		},
-		WatchFunc: func(options kapi.ListOptions) (watch.Interface, error) {
-			return factory.Client.DeploymentConfigs(kapi.NamespaceAll).Watch(options)
-		},
-	}
-	store := cache.NewStore(cache.MetaNamespaceKeyFunc)
-	cache.NewReflector(deploymentConfigLW, &deployapi.DeploymentConfig{}, store, 2*time.Minute).Run()
-
-	changeController := &ImageChangeController{
-		listDeploymentConfigs: func() ([]*deployapi.DeploymentConfig, error) {
-			configs := []*deployapi.DeploymentConfig{}
-			objs := store.List()
-			for _, obj := range objs {
-				configs = append(configs, obj.(*deployapi.DeploymentConfig))
-			}
-			return configs, nil
-		},
-		client: factory.Client,
-	}
-
-	return &controller.RetryController{
-		Queue: queue,
-		RetryManager: controller.NewQueueRetryManager(
-			queue,
-			cache.MetaNamespaceKeyFunc,
-			func(obj interface{}, err error, retries controller.Retry) bool {
-				utilruntime.HandleError(err)
-				if _, isFatal := err.(fatalError); isFatal {
-					return false
-				}
-				if retries.Count > 2 {
-					return false
-				}
-				return true
-			},
-			flowcontrol.NewTokenBucketRateLimiter(1, 10),
-		),
-		Handle: func(obj interface{}) error {
-			repo := obj.(*imageapi.ImageStream)
-			return changeController.Handle(repo)
-		},
+func (c *ImageChangeController) work() bool {
+	key, quit := c.queue.Get()
+	if quit {
+		return true
 	}
+	defer c.queue.Done(key)
+
+	stream, err := c.getByKey(key.(string))
+	if err != nil {
+		glog.Error(err.Error())
+	}
+
+	if stream == nil {
+		return false
+	}
+
+	if err := c.Handle(stream); err != nil {
+		utilruntime.HandleError(err)
```

Comment:
for the future, it would be good if HandleError can accept something like "source" or "prefix" as it is handing a lot of errors atm. and it makes them harder to debug from the log ;-)